### PR TITLE
SONARJAVA-1961 S2095 should rely on flow reporting rather than syntax node in constraint

### PR DESCRIPTION
**Pull request by saberduck**
S2095 should rely on flow reporting rather than syntax node in constraint

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java [hunk 1]
```diff
@@ -72,15 +74,19 @@ public void scanFile(JavaFileScannerContext context) {
     issues.clear();
   }
 
-  public void reportIssue(Tree tree, String message, Set<List<JavaFileScannerContext.Location>> secondary) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java]: I must say that I don't see the point of this refactoring here : if it is a matter of style, well : what is the added value of this effort  ? if there is something else : where is the test ?

File: java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java [hunk 2]
```diff
@@ -131,7 +137,7 @@ public void interruptedExecution(CheckerContext context) {
     public SEIssue(Tree tree, String message, Set<List<JavaFileScannerContext.Location>> flows) {
       this.tree = tree;
       this.message = message;
-      this.flows = flows;
+      this.flows = new HashSet<>(flows);
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java]: AFAI understand : this line is the sole purpose of this commit. If we need to be defensive it could be great to see a test about it.

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 3]
```diff
@@ -92,19 +94,23 @@ public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode)
...
-      }
+    ExplodedGraph.Node node = context.getNode();
+    context.getState().getValuesWithConstraints(Status.OPENED)
+      .forEach((sv, c) -> processUnclosedSymbolicValue(node, sv));
+  }
+
+  private void processUnclosedSymbolicValue(ExplodedGraph.Node node, SymbolicValue sv) {
+    List<JavaFileScannerContext.Location> flow = SECheck.flow(node, sv);
+    flow.stream()
+      .filter(loc -> loc.syntaxNode.is(Tree.Kind.NEW_CLASS) || loc.syntaxNode.is(Tree.Kind.METHOD_INVOCATION))
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: no need for the or, is has multiple arity.

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 4]
```diff
@@ -92,19 +94,23 @@ public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode)
...
+      .forEach(loc -> reportIssue(loc.syntaxNode, String.format("Close this \"%s\".", name(loc.syntaxNode)), ImmutableSet.of(flow)));
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: Something is odd here : for each element of the flow which is new class or method invocation you report an issue with the whole flow as flow for the issue ? 
This tends to demonstrate we probably need something to test the flow thoroughly  in the file.

File: java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java [hunk 5]
```diff
@@ -72,15 +74,19 @@ public void scanFile(JavaFileScannerContext context) {
...
-  public void reportIssue(Tree tree, String message, Set<List<JavaFileScannerContext.Location>> secondary) {
-    issues.add(issues.stream()
-      .filter(seIssue -> seIssue.tree.equals(tree))
-      .findFirst()
-      .map(seIssue -> {
-        seIssue.flows.addAll(secondary);
-        return seIssue;
-      })
-      .orElse(new SEIssue(tree, message, secondary)));
+  public void reportIssue(Tree tree, String message, Set<List<JavaFileScannerContext.Location>> flows) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java]: Kudos for renaming secondary into flows.

File: java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java (hunk 2 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/checks/SECheck.java]: It is tested implicitly by using ImmutableSet when constructing the issue. Will be nice to have explicit test, but not worth the effort to test inner private class

**Comment by saberduck** [comment]
I had to rewrite a history, so I couldn't do incremental commit

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java [hunk 6]
```diff
@@ -70,7 +69,7 @@ public ProgramPoint(CFG.Block block, int i) {
 
     @Override
     public int hashCode() {
-      if(hashcode == 0) {
+      if (hashcode == 0) {
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java]: No need for refactoring but those cosmetics changes are cool, but should not be part of such a commit. Makes the diff harder to read for reviewer.

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 7]
```diff
@@ -92,19 +94,24 @@ public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode)
...
-      }
+    ExplodedGraph.Node node = context.getNode();
+    context.getState().getValuesWithConstraints(Status.OPENED)
+      .forEach((sv, c) -> processUnclosedSymbolicValue(node, sv));
+  }
+
+  private void processUnclosedSymbolicValue(ExplodedGraph.Node node, SymbolicValue sv) {
+    List<JavaFileScannerContext.Location> flow = SECheck.flow(node, sv,
+      constraint -> constraint instanceof ObjectConstraint && ((ObjectConstraint) constraint).hasStatus(Status.OPENED));
+    flow.stream()
```
Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: You are still passing the flow on which you stream when you report the issue.

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 8]
```diff
@@ -19,15 +19,18 @@
  */
 package org.sonar.java.se.checks;
 
+import com.google.common.collect.ImmutableSet;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: MINOR Remove this unused import 'com.google.common.collect.ImmutableSet'. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MINOR 1 minor

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java (hunk 6 above)
Comment (saberduck) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java]: Sorry for this, I will try to separate code reformatting from actual changes. I considered this minor enough, so I didn't bother splitting it. Not sure if you are aware, but you can add `?w=1` to the URL to make diff ignore whitespace (see here https://github.com/blog/967-github-secrets )

Comment (benzonico) [inline java-frontend/src/main/java/org/sonar/java/se/ExplodedGraph.java]: Thanks for the parameter, useful indeed. But as discussed a bit ouside of the point : this, ideally,  should not be part of this commit.
